src/grids.cpp

```diff
@@ -2644,4 +2773,612 @@ ListOfGenericGrids proj_generic_grid_init(PJ *P, const char *gridkey) {
     return grids;
 }
 
+// ---------------------------------------------------------------------------
+
+static const HorizontalShiftGrid *
+findGrid(const ListOfHGrids &grids, const PJ_LP &input,
+         HorizontalShiftGridSet *&gridSetOut) {
+    for (const auto &gridset : grids) {
+        auto grid = gridset->gridAt(input.lam, input.phi);
+        if (grid) {
+            gridSetOut = gridset.get();
+            return grid;
+        }
+    }
+    return nullptr;
+}
+
+// ---------------------------------------------------------------------------
+
+static ListOfHGrids getListOfGridSets(PJ_CONTEXT *ctx, const char *grids) {
+    ListOfHGrids list;
+    auto listOfGrids = internal::split(std::string(grids), ',');
+    for (const auto &grid : listOfGrids) {
+        const char *gridname = grid.c_str();
+        bool canFail = false;
+        if (gridname[0] == '@') {
+            canFail = true;
+            gridname++;
+        }
+        auto gridSet = HorizontalShiftGridSet::open(ctx, gridname);
+        if (!gridSet) {
+            if (!canFail) {
+                pj_ctx_set_errno(ctx, PJD_ERR_FAILED_TO_LOAD_GRID);
+                return {};
+            }
+        } else {
+            list.emplace_back(std::move(gridSet));
+        }
+    }
+    return list;
+}
+
+/**********************************************/
+ListOfHGrids pj_hgrid_init(PJ *P, const char *gridkey) {
+    /**********************************************
+
+      Initizalize and populate list of horizontal
+      grids.
+
+        Takes a PJ-object and the plus-parameter
+        name that is used in the proj-string to
+        specify the grids to load, e.g. "+grids".
+        The + should be left out here.
+
+        Returns the number of loaded grids.
+
+    ***********************************************/
+
+    std::string key("s");
+    key += gridkey;
+    const char *grids = pj_param(P->ctx, P->params, key.c_str()).s;
+    if (grids == nullptr)
+        return {};
+
+    return getListOfGridSets(P->ctx, grids);
+}
+
+// ---------------------------------------------------------------------------
+
+typedef struct { pj_int32 lam, phi; } ILP;
+
+static PJ_LP pj_hgrid_interpolate(PJ_LP t, const HorizontalShiftGrid *grid,
+                                  bool compensateNTConvention) {
```

Comment:
A description of what this interpolation method is, how it was constructed, a reference to a paper? that states this is a good method to use, etc., would be helpful for posterity.

Reply:
comment added